csrc/device_lower/validation.cpp

```diff
@@ -1326,4 +1326,79 @@ void validate1dTmaLoad(Fusion* fusion) {
   }
 }
 
+void validateScatter(Fusion* fusion) {
+  for (auto sop : ir_utils::getOpsOfType<ScatterOp>(fusion)) {
+    auto in_tv = sop->in()->as<TensorView>();
+    auto out_tv = sop->out()->as<TensorView>();
+
+    // TensorIndexer currently only supports scatter with 1D tensors
+    // due to the non-exactness of non-indexed IDs.
+    NVF_ERROR_EQ(
+        out_tv->getLogicalDomain().size(),
+        1,
+        "Scatter with multi-dimensional tensors is not yet supported: ",
+        sop->toString());
+
+    // Scatter is implemented as an in-place op. To lower it safely, it
+    // needs to be able to alias each other. Here are the conditions to
+    // make sure they are valid input and output tensors.
+
+    NVF_ERROR_EQ(
+        in_tv->uses().size(),
+        1,
+        "Scatter input can only be used by the scatter op: ",
+        toDelimitedString(in_tv->uses()));
+
+    NVF_ERROR_EQ(in_tv->getMemoryType(), out_tv->getMemoryType());
+    NVF_ERROR_EQ(in_tv->getDeviceMesh(), out_tv->getDeviceMesh());
+
+    // Assumes both input and output have set allocation domains
+    NVF_ERROR(
+        in_tv->getMemoryType() == MemoryType::Global || in_tv->hasAllocation(),
+        "Non-global scatter input must have an allocation domain");
+    NVF_ERROR(
+        out_tv->getMemoryType() == MemoryType::Global ||
+            out_tv->hasAllocation(),
+        "Non-global scatter output must have an allocation domain");
+
+    auto is_exact_mapped = [](const std::vector<IterDomain*>& ids1,
+                              const std::vector<IterDomain*>& ids2) -> bool {
+      const auto& exact_graph =
+          GpuLower::current()->idModel().idGraph(IdMappingMode::EXACT);
+
+      if (ids1.size() != ids2.size()) {
+        return false;
+      }
+
+      for (const auto i : arange(ids1.size())) {
+        if (!exact_graph.disjointValSets().strictAreMapped(ids1[i], ids2[i])) {
+          return false;
+        }
+      }
+
+      return true;
+    };
+
+    NVF_ERROR(
+        is_exact_mapped(
+            in_tv->getAllocationDomain(), out_tv->getAllocationDomain()),
+        "Scatter input and output must have equivalent allocation domains");
+
+    // Fusion input as scatter input is not yet supported
+    NVF_ERROR(
+        !in_tv->isFusionInput(),
```

Comment:
should we check that `in_tv` is not a FusionOutput?

IIUC, the reaon that we exclude fusion input is to avoid altering the semantics when we change scatter into an inplace operation.


The made-up scenario I have in mind is a case of cascading scatter.

```
fusion( in_tv0) {
  tv1 = op(in_tv0)
  tv2 = scatter(tv1, ...)
  tv3 = scatter(tv2, ...)
  return tv2, tv3
}
```

I think our existing logic is going to alias everything together (tv1, tv2, tv3). which means we'll produce two output with the same value and reuse buffer.

But that's not the original program any more.

Reply:
^^^ Not totally sure if this is a legit argument. I'm not fully grasping what's the protocol we are establishing with our scatter API.

Reply:
That's a good point. Yes, using a fusion output as the input of a scatter is not allowed.